PWGHF/D2H/Macros/efficiency.py

```diff
@@ -0,0 +1,389 @@
+        x_list = x.split(',')
+        for i, element in enumerate(x_list):
+            x_list[i] = element.strip() # remove possible whitespaces
+        return x_list
+
+    return x
+
+def enforce_trailing_slash(path: str) -> str:
+    """
+    Helper method to enforce '/' at the and of directory name
+
+    Parameters
+    ----------
+    - path: some path
+
+    Returns
+    ----------
+    - path with a trailing slash at the end if it was not there yet
+    """
+
+    if path[-1] != '/':
+        path += '/'
+
+    return path
+
+def save_canvas(
+    canvas: TCanvas,
+    out_dir: str,
+    name_file: str,
+    extension: list) -> None:
+    """
+    Save canvas in formats chosen by extension
+
+    Parameters
+    ----------
+    - canvas: a TCanvas instance
+    - out_dir: output directory where canvas will be saved
+    - name_file: name of the output file
+    - extension: file format
+    """
+
+    for ext in extension:
+        canvas.SaveAs(out_dir + name_file + '.' + ext)
+
+def compute_efficiency(
+    h_rec: Union[TH1, TH2],
+    h_gen: Union[TH1, TH2],
+    rapidity_cut: float = None,
+    axis_rapidity: str = 'Y') -> TEfficiency:
+    """
+    Helper method to compute the efficiency as function of the feature in axis.
+
+    Parameters
+    ----------
+    - h_rec: histogram with the reconstructed information
+    - h_gen: histogram with the generated information
+    - rapidity_cut: applies the selection |y| < y_cut to the efficiency
+    - axis_rapidity: histogram axis containting rapidity
+
+    Returns
+    ----------
+    - efficiency: a TEfficiency instance
+    """
+
+    epsilon = 0.0001
+
+    is_reco_2d, is_gen_2d = False, False
+    if h_rec.GetDimension() == 2:
+        is_reco_2d = True
+    if h_gen.GetDimension() == 2:
+        is_gen_2d = True
+
+    if is_reco_2d:
+        if axis_rapidity == 'Y':
+            h_rec = h_rec.ProjectionX()
+        else:
+            h_rec = h_rec.ProjectionY()
+
+    if is_gen_2d:
+        if rapidity_cut is not None:
+            h_gen.GetYaxis().SetRangeUser(-1.0 * rapidity_cut + epsilon, rapidity_cut - epsilon)
+        if axis_rapidity == 'Y':
+            h_gen = h_gen.ProjectionX()
+        else:
+            h_gen = h_gen.ProjectionY()
+
+    efficiency = TEfficiency(h_rec, h_gen)
+
+    return efficiency
+
+def get_th1_from_tefficiency(
+    teff: TEfficiency,
+    h_eff: TH1) -> TH1:
+    """
+    Helper method to convert TEfficiency object to TH1
+
+    Parameters
+    ----------
+    - teff: TEfficiency instance
+    - h_eff: clone of histogram with reconstructed information
+
+    Returns
+    ----------
+    - h_eff: histogram containing teff information
+    """
+
+    n_bins = h_eff.GetXaxis().GetNbins()
+    for i_bin in range(1, n_bins+1):
+        h_eff.SetBinContent(i_bin, teff.GetEfficiency(i_bin))
+        # TH1 can't handle asymmetric errors so we take the max
+        err_max = max(teff.GetEfficiencyErrorLow(i_bin), teff.GetEfficiencyErrorUp(i_bin))
```

Comment:
I'm wondering in which case it can happen that the efficiency has an asymmetric error. If I am not wrong, the binomial uncertainty is always symmetric right?
Just to be sure that I'm not overlooking anything, the code looks fine.

Reply:
Hi Mattia, thanks for the reply. Indeed, I must say I could not think of a case where this would happen, but I put it as extra safety after reading about it in a ROOT forum. But, in our use case, I don't think that we truly need it, if I am not misssing something.

Reply:
> Hi Mattia, thanks for the reply. Indeed, I must say I could not think of a case where this would happen, but I put it as extra safety after reading about it in a ROOT forum. But, in our use case, I don't think that we truly need it, if I am not misssing something.

Hi @AlexBigO , exactly! This is my point.
What about adding a warning when this happens? Something like:

`if abs(teff.GetEfficiencyErrorLow(i_bin) - teff.GetEfficiencyErrorUp(i_bin))/teff.GetEfficiencyErrorLow(i_bin) > 1e-05:
    print a warning`

Something similar to this. What do you think?

Reply:
Ah, this is better indeed! Thanks, I implemented it, and will push it along with future comments.

Reply:
thank you!